File: CoMD/python/cell/halo.py

import numpy as np

class Halo:
    HALO_X_MINUS = 0
    HALO_X_PLUS = 1
    HALO_Y_MINUS = 2
    HALO_Y_PLUS = 3
    HALO_Z_MINUS = 4
    HALO_Z_PLUS = 5

    HALO_X_AXIS = 0
    HALO_Y_AXIS = 1
    HALO_Z_AXIS = 2

    def __init__(self, boxes):
        self.bufCapacity = 0
        self.nCells = np.zeros(6, dtype=np.int)
        self.pbcFactor = np.zeros((6,3))

        self.nCells[self.HALO_X_MINUS] = 2*(boxes.gridSize[1] + 2) * (boxes.gridSize[2] + 2)
        self.nCells[self.HALO_Y_MINUS] = 2*(boxes.gridSize[0] + 2) * (boxes.gridSize[2] + 2)
        self.nCells[self.HALO_Z_MINUS] = 2*(boxes.gridSize[0] + 2) * (boxes.gridSize[1] + 2)
        self.nCells[self.HALO_X_PLUS] = self.nCells[self.HALO_X_MINUS]
        self.nCells[self.HALO_Y_PLUS] = self.nCells[self.HALO_Y_MINUS]
        self.nCells[self.HALO_Z_PLUS] = self.nCells[self.HALO_Z_MINUS]

        self.cellList = []
        for ii in range(6):
            self.cellList.append(self.mkAtomCellList(boxes, ii))

        self.pbcFactor[self.HALO_X_MINUS, self.HALO_X_AXIS] = 1.0
        self.pbcFactor[self.HALO_X_PLUS, self.HALO_X_AXIS] = -1.0
        self.pbcFactor[self.HALO_Y_MINUS, self.HALO_Y_AXIS] = 1.0
        self.pbcFactor[self.HALO_Y_PLUS, self.HALO_Y_AXIS] = -1.0
        self.pbcFactor[self.HALO_Z_MINUS, self.HALO_Z_AXIS] = 1.0
        self.pbcFactor[self.HALO_Z_PLUS, self.HALO_Z_AXIS] = -1.0


    def haloExchange(self, sim):
        for iAxis in range(3):
            self.exchangeData(iAxis, sim)

    def exchangeData(self, iAxis, sim):
        faceM = 2*iAxis
        faceP = faceM + 1
        bufM = self.loadAtomsBuffer(sim, faceM)
        bufP = self.loadAtomsBuffer(sim, faceP)
        self.unloadAtomsBuffer(sim, faceM, bufP)
        self.unloadAtomsBuffer(sim, faceP, bufM)

    def loadAtomsBuffer(self, sim, face):
        shift = self.pbcFactor[face,:] * sim.atoms.bounds
        nCells = self.nCells[face]
        cellList = self.cellList[face]
        buf = []
        for iCell in range(nCells):
            iBox = cellList[iCell]
            for ii in range(sim.boxes.nAtoms[iBox]):
                iOff = iBox*sim.boxes.MAXATOMS + ii
                entry = {}
                entry['gid'] = sim.atoms.gid[iOff]
                entry['type'] = sim.atoms.species[iOff]
                entry['rx'] = sim.atoms.r[iOff][0] + shift[0]
                entry['ry'] = sim.atoms.r[iOff][1] + shift[1]
                entry['rz'] = sim.atoms.r[iOff][2] + shift[2]
                entry['px'] = sim.atoms.p[iOff][0]
                entry['py'] = sim.atoms.p[iOff][1]
                entry['pz'] = sim.atoms.p[iOff][2]
                buf.append(entry)

        return buf

    def unloadAtomsBuffer(self, sim, face, buf):
        for entry in buf:
            gid = entry['gid']
            species = entry['type']
            rx = entry['rx']
            ry = entry['ry']
            rz = entry['rz']
            px = entry['px']
            py = entry['py']
            pz = entry['pz']
            sim.boxes.putAtomInBox(sim.atoms, gid, species, rx, ry, rz, px, py, pz)

    def mkAtomCellList(self, boxes, iFace):
        xBegin = -1
        xEnd = boxes.gridSize[0] + 1
        yBegin = -1
        yEnd = boxes.gridSize[1] + 1
        zBegin = -1
        zEnd = boxes.gridSize[2] + 1

        if iFace == self.HALO_X_MINUS: xEnd = xBegin + 2
        if iFace == self.HALO_X_PLUS:  xBegin = xEnd - 2
        if iFace == self.HALO_Y_MINUS: yEnd = yBegin + 2
        if iFace == self.HALO_Y_PLUS:  yBegin = yEnd - 2
        if iFace == self.HALO_Z_MINUS: zEnd = zBegin + 2
        if iFace == self.HALO_Z_PLUS:  zBegin = zEnd - 2

        cells = []
        for ix in range(xBegin,xEnd):
            for iy in range(yBegin,yEnd):
                for iz in range(zBegin,zEnd):
                    cells.append(boxes.getBoxFromTuple(ix,iy,iz))
        return cells

File: CoMD/julia/cell/halo.jl

include("constants.jl")

HALO_X_MINUS = 1
HALO_X_PLUS = 2
HALO_Y_MINUS = 3
HALO_Y_PLUS = 4
HALO_Z_MINUS = 5
HALO_Z_PLUS = 6

HALO_X_AXIS = 1
HALO_Y_AXIS = 2
HALO_Z_AXIS = 3

mutable struct HaloExchange
    nCells::Array{Int, 1}
    pbcFactor::Array{Float, 2}
    cellList::Array{Array{Int, 1}, 1}
end

function initHalo(boxes)
    nCells = zeros(Int, 6)
    pbcFactor = zeros(Float, 6, 3)

    gs = boxes.gridSize

    nCells[HALO_X_MINUS] = 2*(gs[2]+2) * (gs[3] + 2)
    nCells[HALO_Y_MINUS] = 2*(gs[1]+2) * (gs[3] + 2)
    nCells[HALO_Z_MINUS] = 2*(gs[1]+2) * (gs[2] + 2)
    nCells[HALO_X_PLUS] = nCells[HALO_X_MINUS]
    nCells[HALO_Y_PLUS] = nCells[HALO_Y_MINUS]
    nCells[HALO_Z_PLUS] = nCells[HALO_Z_MINUS]

    cellList = Array{Any}(undef, 6)
    for ii in 1:6
        cellList[ii] = mkAtomCellList(boxes, ii, nCells[ii])
    end

    pbcFactor[HALO_X_MINUS, HALO_X_AXIS] = 1.0
    pbcFactor[HALO_X_PLUS, HALO_X_AXIS] = -1.0
    pbcFactor[HALO_Y_MINUS, HALO_Y_AXIS] = 1.0
    pbcFactor[HALO_Y_PLUS, HALO_Y_AXIS] = -1.0
    pbcFactor[HALO_Z_MINUS, HALO_Z_AXIS] = 1.0
    pbcFactor[HALO_Z_PLUS, HALO_Z_AXIS] = -1.0

    HaloExchange(nCells, pbcFactor, cellList)
end

function mkAtomCellList(boxes, iFace, nCell)
    xBegin = -1
    xEnd = boxes.gridSize[1] + 1
    yBegin = -1
    yEnd = boxes.gridSize[2] + 1
    zBegin = -1
    zEnd = boxes.gridSize[3] + 1

    if iFace == HALO_X_MINUS
        xEnd = xBegin + 2
    end
    if iFace == HALO_X_PLUS
        xBegin = xEnd - 2
    end
    if iFace == HALO_Y_MINUS
        yEnd = yBegin + 2
    end
    if iFace == HALO_Y_PLUS
        yBegin = yEnd - 2
    end
    if iFace == HALO_Z_MINUS
        zEnd = zBegin + 2
    end
    if iFace == HALO_Z_PLUS
        zBegin = zEnd - 2
    end

    list = Array{Int}(undef,nCell)

    ii = 1
    for ix in xBegin:xEnd-1
        for iy in yBegin:yEnd-1
            for iz in zBegin:zEnd-1
                iBox = getBoxFromTuple(boxes, ix, iy, iz)
                list[ii] = iBox
                ii += 1
            end
        end
    end

    return list
end

function haloExchange(halo, atoms, boxes)
    for iAxis in 1:3
        exchangeData(halo, iAxis, atoms, boxes)
    end
end

function exchangeData(halo, iAxis, atoms, boxes)
    faceM = 2*(iAxis-1) + 1
    faceP = faceM + 1
    bufM = loadAtomsBuffer(halo, atoms, boxes, faceM)
    bufP = loadAtomsBuffer(halo, atoms, boxes, faceP)
    unloadAtomsBuffer(halo, atoms, boxes, faceM, bufP)
    unloadAtomsBuffer(halo, atoms, boxes, faceP, bufM)
end

mutable struct TransferBuffer
    gid::Int
    species::Int
    r::Array{Float}
    p::Array{Float}
end

function loadAtomsBuffer(halo, atoms, boxes, face)
    shift = halo.pbcFactor[face,:] .* atoms.bounds

    nCells = halo.nCells[face]
    cellList = halo.cellList[face]
    size = 0
    for iCell in 1:nCells
        iBox = cellList[iCell]
        for ii in 1:boxes.nAtoms[iBox]
            size += 1
        end
    end

    buf = Array{TransferBuffer}(undef, size)
    idx = 1
    for iCell in 1:nCells
        iBox = cellList[iCell]
        for ii in 1:boxes.nAtoms[iBox]
            iOff = (iBox-1)*boxes.MAXATOMS + ii
            pos = atoms.r[:, iOff] + shift
            entry = TransferBuffer(atoms.gid[iOff], atoms.species[iOff], pos, atoms.p[:,iOff])
            buf[idx] = entry
            idx += 1
        end
    end
    return buf
end

function unloadAtomsBuffer(halo, atoms, boxes, face, buf)
    for entry in buf
        putAtomInBox(boxes, atoms, entry.gid, entry.species, entry.r, entry.p)
    end
end


